### PR TITLE
Consume new `References3.AddFiles` API

### PR DESCRIPTION
**Pull request by tmeschter**
## Bug
Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/572174
Regression: No
If Regression then when did it last work:  N/A
If Regression then how are we preventing it in future:   N/A

## Fix
Details:

This change adds basic usage of the new `References3.AddFiles` API when
adding references in a packages.config-based project. The current API
used for this purposes, `References.Add`, can trigger multiple
design-time builds per call and may not set all the right metadata on
the resulting `<Reference>` element in the project file. This can
require further calls to update that metadata, which can in turn trigger
further design-time builds.

`AddFiles` on the other hand creates `<Reference>` items with all the
proper metadata--notable, the `<Private>` metadata is set appropriately,
and it specifies an explicit version of the given assembly. And since it
is explicitly about adding files as references and does not need to
support short or full assembly names it can skip several of the
design-time builds run by `Add`. Use of `AddFiles` instead of `Add`
eliminates up to 75% of the design-time builds incurred while adding a
new package.

For the time being, if `AddFiles` fails we will fall back to calling
`Add`. This way adding a new package is not completely broken in case
there are bugs in the `AddFiles` implementation. If the new API proves
reliable, then the fallback can be removed later.

`AddFiles` can also take multiple file paths, and resolve all the
assemblies with a single design-time build. However, using the API in
that manner will require some refactoring of the client code; it is an
option for the future but is not pursued here.

## Testing/Validation
Tests Added: No  
Reason for not adding tests:  Change concerns integration with Visual Studio
Validation done:  Manually verified that with the right changes in place to both the NuGet client and the underlying project system adding new packages works end-to-end.

### CODE REVIEW

**Comment by tmeschter** [comment]
Note that this change will not build under the new `VSLangProj157` package has been uploaded to MyGet or NuGet; as such this should not be merged until after that is done.

**Comment by rohit21agrawal** [comment]
@tmeschter let me know when you are ready to publish the package, and we will run this through our CI

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:738]: should we differentiate between NPE vs other exceptions? like if it fails to type cast into `Reference3`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:816]: Shouldn't we now remove setting CopyLocal explicitly? Look at line no 840

Comment (tmeschter) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:816]: If we fall back to `References.Add` we still need to set CopyLocal explicitly.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs
Comment (tmeschter) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:738]: Hmm. What if we reported a telemetry event if the type cast fails? That should only happen if there's a setup authoring issue (e.g. the type library wasn't registered correctly or something like that) or we're encountering a project system that implements `References` but not `References3`. Either way we would want to know.

Does that sound reasonable?

Comment (jainaashish) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:738]: yeah that sounds good.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs
Comment (jainaashish) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:816]: hmm i see, will there be any perf impact of setting this CopyLocal for `Reference3`?

Comment (tmeschter) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:816]: No, since `TrySetCopyLocal` first checks the `CopyLocal` value, and only sets it if it is not `true`. When using `References3.AddFiles` it will always be `true`.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs
Comment (tmeschter) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:738]: @jainaashish I've added a very simple telemetry event.

**Review by jainaashish — APPROVED** [review]
Looks good, thanks Tom for the PR.

**Comment by jainaashish** [comment]
@tmeschter is the package published?

**Comment by tmeschter** [comment]
@jainaashish Not yet. The PR with the project system implementation (including the package .nuspec) has been approved as of today, but I'm waiting until Monday to merge it in on the off chance it breaks something. :-) Once it has gone in I'll pick up the package from the next build.

**Comment by zhili1208** [comment]
@tmeschter does this PR have any VS version requirement? it can't find "VSLangProj157, Version=15.7.0.0" in our EndToEnd test.

**Comment by rohit21agrawal** [comment]
@zhili1208 you might want to take a look at how the EndToEnd.zip is created, maybe you can workaround this by including the above DLL directly into the package.

**Comment by zhili1208** [comment]
@rohit21agrawal it's not just EndToEnd test, apex test also failed, looks like API only works in 15.7, just manually tried 15.6, didn't work. we need to update the VS in our CI agent

**Comment by zhili1208** [comment]
@tmeschter is there any way to get 15.6 works with this API?

**Comment by tmeschter** [comment]
@zhili1208 No, this depends on a new API that isn't available prior to 15.7.

**Comment by zhili1208** [comment]
merged those two commits in https://github.com/NuGet/NuGet.Client/pull/2133

**Pull request closed without merging**
